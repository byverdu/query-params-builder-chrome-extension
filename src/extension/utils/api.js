export const OPTIONS_ITEM = 'QueryParamsBuilderOptions';
export const TABS_ITEM = 'QueryParamsBuilderTab';
export const GET_STORAGE = 'GET_STORAGE';
export const SET_STORAGE = 'SET_STORAGE';
export const GET_CURRENT_TAB = 'GET_CURRENT_TAB';
export const UPDATE_URL_TAB = 'UPDATE_URL_TAB';
export const REMOVE_ALL_STORAGE = 'REMOVE_ALL_STORAGE';

/**
 * @type {import('../types/index.js').API}
 */
export const extensionApi = {
  setStorage: async (key, value) =>
    await chrome.storage.sync.set({
      [key]: value,
    }),
  setStorageSync: (key, value) =>
    chrome.storage.sync.set({
      [key]: value,
    }),
  getStorage: async key => await chrome.storage.sync.get(key),
  getStorageSync: (key, callback) => chrome.storage.sync.get(key, callback),
  sendMessage: async ({ type, payload }) =>
    await chrome.runtime.sendMessage({
      type,
      payload,
    }),
  onRemovedTab: callback => chrome.tabs.onRemoved.addListener(callback),
  onMessage: callback => chrome.runtime.onMessage.addListener(callback),
  getCurrentTab: async () =>
    await chrome.tabs.query({ active: true, currentWindow: true }),
  updateUrlTab: async url => await chrome.tabs.update({ url }),
  removeStorage: async key => await chrome.storage.sync.remove(key),
};
