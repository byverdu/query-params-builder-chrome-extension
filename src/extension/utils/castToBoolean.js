export function castToBoolean(value) {
  try {
    const parsed = JSON.parse(value);
    return parsed;
  } catch {
    return null;
  }
}
