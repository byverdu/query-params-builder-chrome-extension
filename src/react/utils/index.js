/// <reference path="../../types/index.d.ts" />

import {
  extensionApi,
  GET_STORAGE,
  OPTIONS_ITEM,
  TABS_ITEM,
} from '../../extension/utils/api.js';

/**
 * @type {API}
 */
const { sendMessage } = extensionApi;

/**
 * @param {React.Dispatch<React.SetStateAction<Toast>>} setToast
 * @param {Error} error
 * @param {UpdateActions | ExtensionActions} type
 */
const sendMessageCatchHandler = (setToast, error, type) => {
  console.error(`QueryParamsBuilder extension ${type}`, String(error));
  setToast({ type: 'danger', text: `${type}: ${error.message}` });
};

/**
 *
 * @param {Promise} promise
 * @param {React.Dispatch<React.SetStateAction<Toast>>} setToast
 * @param {string} text
 * @param {UpdateActions | ExtensionActions} type
 */
const sendMessageAsyncHandler = (promise, setToast, text, type) => {
  promise
    .then(() => {
      setToast({ type: 'success', text });
    })
    .catch(error => {
      console.error(`QueryParamsBuilder extension ${type}`, String(error));
      setToast({
        type: 'danger',
        text: `${type}: ${error.message}`,
      });
    });
};

/**
 *
 * @param {string | null} tabUrl
 * @returns {Promise<OptionsExtensionProps[] | BaseExtensionProps[]>}
 */
const fetchTabStorage = async tabUrl => {
  /**
   * @type {{[key: string]: OptionsExtensionProps[]}}
   */
  const tabs = await sendMessage({
    type: GET_STORAGE,
    payload: { value: TABS_ITEM },
  });

  /**
   * @type {BaseExtensionProps[]}
   */
  const options = await sendMessage({
    type: GET_STORAGE,
    payload: { value: OPTIONS_ITEM },
  });

  const validOptions = options && Array.isArray(options) ? options : [];
  const savedTabInfo = tabs && tabs[tabUrl];

  if (!savedTabInfo) {
    return validOptions;
  }

  const validTabsInfo =
    savedTabInfo && Array.isArray(savedTabInfo) ? savedTabInfo : [];

  for (const option of validOptions) {
    if (!validTabsInfo.find(item => item.id === option.id)) {
      validTabsInfo.push(option);
    }
  }

  return validTabsInfo;
};

/**
 * @param {BaseExtensionProps[]} prevState
 * @param {BaseExtensionProps} newValue
 * @returns
 */
const updateState = (prevState, newValue) => {
  if (Array.isArray(prevState)) {
    return [...prevState, newValue];
  }

  return prevState;
};

/**
 * @param {BaseExtensionProps[]} prevState
 * @param {string} id
 * @returns
 */
const removeItemFromState = (prevState, id) => {
  if (Array.isArray(prevState) && typeof id === 'string') {
    return prevState.filter(item => item.id !== id);
  }

  return prevState;
};

/**
 * @param {BaseExtensionProps[]} prevState
 * @param {{id: string, key: string, value: string}} id
 * @returns
 */
const editItemFromState = (prevState, { id, key, value }) => {
  if (Array.isArray(prevState) && id && typeof id === 'string') {
    return prevState.map(item => {
      if (item.id === id) {
        item[key] = value;
      }

      return item;
    });
  }

  return prevState;
};

/**
 *
 * @param {"options" | "popup"} type
 * @param {HTMLFormControlsCollection} elements
 */
const getNewItemToSave = (type, elements) => {
  const canDeleteFromPopup = type === 'popup';
  const newItem = {
    checked: false,
    canDeleteFromPopup,
    id: crypto.randomUUID(),
  };

  for (const elem of elements) {
    if (elem.nodeName === 'INPUT') {
      newItem[elem.id] = elem.value;

      // remove old value
      elem.value = '';
    }
  }

  return newItem;
};

export {
  sendMessageCatchHandler,
  fetchTabStorage,
  sendMessageAsyncHandler,
  updateState,
  getNewItemToSave,
  removeItemFromState,
  editItemFromState,
};
