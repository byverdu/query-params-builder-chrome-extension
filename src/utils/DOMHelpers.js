function tableRowBuilder(id, bundleName, urlParamKey) {
  return `
  <tr id="${id}">
  <td>${bundleName}</td>
  <td>${urlParamKey}</td>
  <td>
    <button data-bundle-id="${id}" class="btn btn-outline-danger delete-bundle">
      Delete
    </button>
  </td>
</tr>
  `;
}

/**
 * @param {import('../types/index.js').ExtensionOptions[]} optionsToBuild
 * @param {HTMLTableSectionElement} tbody
 */
export function optionsToTableDefinitionBuilder(optionsToBuild, tbody) {
  for (const option of optionsToBuild) {
    const tr = tableRowBuilder(
      option.id,
      option.bundleName,
      option.urlParamKey
    );

    tbody.innerHTML += tr;
  }
}

export function listItemBuilder({
  id,
  bundleName,
  urlParamKey,
  checked,
  urlParamValue,
  canDeleteFromPopup,
}) {
  return `
  <li class="list-group-item">
    <input
      class="form-check-input me-1"
      type="checkbox"
      value="${urlParamKey}"
      data-url-param-key="${urlParamKey}"
      data-bundle-name="${bundleName}"
      id="${id}"
      ${(checked && 'checked') || ''}
      data-can-delete-from-popup="${canDeleteFromPopup}"
    >
    <label class="form-check-label" for="${id}">${bundleName}</label>
    <input
      type="text"
      class="form-control"
      data-id="${id}"
      placeholder="${urlParamKey} value"
      value="${(urlParamValue && urlParamValue) || ''}"
    />
    ${
      (canDeleteFromPopup &&
        '<button class="btn btn-danger delete-new-item">-</button>') ||
      ''
    }
  </li>
`;
}

/**
 * @param {import('../types/index.js').ExtensionOptions[]} optionsToBuild
 * @param {HTMLElement} target
 * @param {boolean} checked
 */
export function popupOptionsBuilder(optionsToBuild) {
  const listGroup = document.querySelector('.list-group');
  let groupedListContent = '';

  for (const {
    id,
    bundleName,
    urlParamKey,
    urlParamValue,
    checked,
    canDeleteFromPopup,
  } of optionsToBuild) {
    groupedListContent += listItemBuilder({
      id,
      bundleName,
      urlParamKey,
      checked,
      urlParamValue,
      canDeleteFromPopup,
    });
  }

  if (listGroup) {
    listGroup.insertAdjacentHTML('beforeend', groupedListContent);
  } else {
    const target = document.getElementById('selected_bundles');
    const newListGroup = document.createElement('ul');
    newListGroup.classList.add('list-group');
    newListGroup.innerHTML = groupedListContent;

    target.insertAdjacentElement('afterbegin', newListGroup);
  }
}

/**
 * @param {import('../types/index.js').Toast} params
 */
export function setToastContent({ toastType, bodyToastText }) {
  let toast, headerToast, bodyToast;

  const optionsPageToastText = {
    success: 'Success!',
    danger: 'Error!',
  };

  const toastElement = document.getElementById('optionsPageToast');
  if (toastElement) {
    toast = new bootstrap.Toast(toastElement, {
      animation: true,
      autohide: true,
      delay: 3000,
    });

    headerToast = toastElement.querySelector('.toast-header .badge');
    bodyToast = toastElement.querySelector('.toast-body');
  }

  if (toast && headerToast && bodyToast) {
    headerToast.textContent = optionsPageToastText[toastType];
    bodyToast.textContent = bodyToastText;

    headerToast.className += ` text-bg-${toastType}`;

    toast.show();
  }
}

export function randomId() {
  return crypto.randomUUID();
}

export function castToBoolean(value) {
  try {
    const parsed = JSON.parse(value);
    return Boolean(parsed);
  } catch {
    return false;
  }
}

export function getCheckboxesValues() {
  const tabInfoToSave = [];

  /**
   * @type HTMLInputElement[]
   */
  const checkboxes = Array.from(
    document.querySelectorAll('input[type="checkbox"]')
  );

  for (const input of checkboxes) {
    const canDeleteFromPopup = castToBoolean(input.dataset.canDeleteFromPopup);
    const bundleId = input.id;
    const urlParamKey = input.value;
    const urlParamValue = document.querySelector(
      `[data-id="${bundleId}"]`
    ).value;
    const bundleName = input.dataset.bundleName;

    tabInfoToSave.push({
      id: bundleId,
      canDeleteFromPopup,
      checked: input.checked,
      urlParamKey,
      bundleName,
      urlParamValue,
    });
  }

  return tabInfoToSave;
}
