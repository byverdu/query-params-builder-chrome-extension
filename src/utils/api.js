const GET_STORAGE = 'GET_STORAGE';
const SET_STORAGE = 'SET_STORAGE';
const GET_CURRENT_TAB = 'GET_CURRENT_TAB';
const UPDATE_URL_TAB = 'UPDATE_URL_TAB';
const REMOVE_ALL_STORAGE = 'REMOVE_ALL_STORAGE';

export const actions = {
  GET_STORAGE,
  SET_STORAGE,
  GET_CURRENT_TAB,
  UPDATE_URL_TAB,
  REMOVE_ALL_STORAGE,
};

export const extensionApi = {
  setStorage: async (key, value) =>
    await chrome.storage.sync.set({
      [key]: value,
    }),
  getStorage: async key => await chrome.storage.sync.get(key),
  sendMessage: async ({ type, payload }) =>
    await chrome.runtime.sendMessage({
      type,
      payload,
    }),

  onMessage: callback => chrome.runtime.onMessage.addListener(callback),
  getCurrentTab: async () =>
    await chrome.tabs.query({ active: true, currentWindow: true }),
  updateUrlTab: async url => await chrome.tabs.update({ url }),
  removeStorage: async key => await chrome.storage.sync.remove(key),
};
