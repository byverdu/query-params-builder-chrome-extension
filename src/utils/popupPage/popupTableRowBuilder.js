/**
 * @typedef {Object} ExtendedProps
 * @property {boolean} withExtraThead
 * @typedef {import('../../types/index.js').ExtensionOptions & ExtendedProps} Props
 */

/**
 * @param {Props} param
 * @returns string
 */
export function popupTableRowBuilder({
  id,
  bundleName,
  urlParamKey,
  checked,
  urlParamValue,
  canDeleteFromPopup,
  withExtraThead,
}) {
  return `
  <tr>
    <td class="td-checkbox">
      <input
        class="form-check-input me-1"
        type="checkbox"
        value="${urlParamKey}"
        data-url-param-key="${urlParamKey}"
        data-bundle-name="${bundleName}"
        id="${id}"
        ${(checked && 'checked') || ''}
        data-can-delete-from-popup="${canDeleteFromPopup}"
      >
      <label class="form-check-label" for="${id}">${bundleName}</label>
    </td>
    <td>
      <input
        type="text"
        class="form-control"
        data-id="${id}"
        placeholder="${urlParamKey} value"
        value="${(urlParamValue && urlParamValue) || ''}"
      />
    </td>
    ${
      (canDeleteFromPopup &&
        '<td><button class="btn btn-danger delete-new-item">-</button></td>') ||
      (withExtraThead && '<td></td>') ||
      ''
    }
  </tr>
`;
}
