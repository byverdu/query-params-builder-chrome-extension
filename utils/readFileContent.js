import { readFileSync } from 'fs';
import { resolve } from 'path';

/**
 *
 * @param {"options" | "popup"} fileName
 * @returns
 */
function readFileContent(fileName) {
  const fileNameMapper = {
    options: 'src/pages/options/index.html',
    popup: 'src/pages/popup/index.html',
  }[fileName];
  const filePath = resolve('.', fileNameMapper);

  return readFileSync(filePath, {
    encoding: 'utf-8',
  });
}

export { readFileContent };
