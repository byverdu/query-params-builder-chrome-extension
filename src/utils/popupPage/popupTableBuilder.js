import { popupTableRowBuilder } from './popupTableRowBuilder.js';

/**
 *
 * @param {HTMLTableElement} table
 */
function appendExtraTableHeader(table) {
  const tableHeader = table.querySelector('thead tr');

  if (!tableHeader) {
    throw new Error('No thead tr selector present on the document');
  }

  tableHeader.insertAdjacentHTML(
    'beforeend',
    '<th id="withExtraThead" scope="col">Actions</th>'
  );
}

/**
 * @param {import('../../types/index.js').ExtensionOptions[]} optionsToBuild
 */
export function popupTableBuilder(optionsToBuild) {
  /**
   * @type HTMLTableElement | null
   */
  const table = document.querySelector('.table');

  if (!table) {
    throw new Error('No .table selector present on the document');
  }

  const tbody = table.querySelector('tbody');
  let tableRows = '';
  const withExtraThead = optionsToBuild.some(
    option => option.canDeleteFromPopup
  );

  if (withExtraThead && !document.getElementById('withExtraThead')) {
    appendExtraTableHeader(table);
  }

  for (const {
    id,
    bundleName,
    urlParamKey,
    urlParamValue,
    checked,
    canDeleteFromPopup,
  } of optionsToBuild) {
    tableRows += popupTableRowBuilder({
      id,
      bundleName,
      urlParamKey,
      checked,
      urlParamValue,
      canDeleteFromPopup,
      withExtraThead,
    });
  }

  if (tbody) {
    tbody.insertAdjacentHTML('beforeend', tableRows);
  }

  if (optionsToBuild.length === 1 && withExtraThead) {
    tbody.querySelectorAll('tr').forEach(elem => {
      const withoutDeleteBtn = elem.querySelectorAll('td').length === 2;
      if (withoutDeleteBtn) {
        elem.insertAdjacentHTML('beforeend', '<td></td>');
      }
    });
  }
}
