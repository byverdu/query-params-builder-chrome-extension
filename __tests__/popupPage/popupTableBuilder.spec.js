import { popupTableBuilder } from '../../src/utils/popupPage/popupTableBuilder';
import { readFileContent } from '../../utils/readFileContent';
import { propsBuilder } from './helpers';

describe('popupTableBuilder', () => {
  beforeEach(() => {
    document.body.innerHTML = readFileContent('popup');
    jest.resetAllMocks();
  });

  it('should be defined', () => {
    expect(popupTableBuilder).toBeInstanceOf(Function);
  });

  it('should throw an error if no table markup is found', () => {
    document.body.innerHTML = '';
    expect(() => popupTableBuilder([propsBuilder()])).toThrowError(
      'No .table selector present on the document'
    );
  });

  it('should not render extra th if an item can not be deleted', () => {
    popupTableBuilder([propsBuilder()]);

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('should render extra th if an item can be deleted', () => {
    popupTableBuilder([propsBuilder({ canDeleteFromPopup: true })]);

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
