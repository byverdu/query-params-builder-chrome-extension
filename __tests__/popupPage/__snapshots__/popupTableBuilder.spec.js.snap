// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`popupTableBuilder should not render extra th if an item can not be deleted 1`] = `
"

  
    <title>Options</title>

    <link rel="shortcut icon" type="image/jpg" href="../../assets/icons/128x128.png">
    <link href="../../libs/bootstrap.min.css" rel="stylesheet">

    <style>
      .wrapper {
        padding-top: 20px;
        box-sizing: content-box;
      }

      body::-webkit-scrollbar {
        width: 5px;
        height: 8px;
        background-color: #aaa;
      }
      body::-webkit-scrollbar-thumb {
        background: #dc3545;
      }
      #content {
        visibility: hidden;
      }

      #popup-spinner {
        position: absolute;
        left: 45%;
        top: 40%;
      }

      .delete-new-item {
        display: block;
        margin: 0 auto;
      }

      #extraParam .form-control {
        width: 100%;
      }

      #extraParam .col-12 {
        margin: 10px auto;
      }

      .card {
        border: 0;
      }

      .actions {
        padding: 12px 0;
        background-color: rgb(0 0 0 / 90%);
      }

      .action-btns {
        width: 50%;
        margin: 0 auto;
        display: block;
      }

      .table {
        visibility: hidden;
      }

      .td-checkbox {
        display: flex;
        height: 54px;
        align-items: center;
      }
      .td-checkbox label {
        margin-left: 12px;
      }

      #withExtraThead {
        display: none;
      }
    </style>
  
  
    <div class="wrapper">
      <div class="container" id="popup-spinner">
        <div>
          <div class="spinner-border text-primary" role="status"></div>
        </div>
      </div>
      <div class="container-fluid" id="content">
        <div class="row"><h3>QueryParamsBuilder</h3></div>
        <div class="row" style="padding: 0 12px">
          <table class="table table-striped table-bordered">
            <thead>
              <tr>
                <th scope="col" style="display: flex; justify-content: space-between">
                  <span>Params to Apply</span>
                  |
                  <label for="check-all">
                    <input id="check-all" class="form-check-input check-all" type="checkbox">
                    Check All
                  </label>
                </th>
                <th scope="col">Url param value</th>
                <th id="withExtraThead" scope="col">Actions</th>
              </tr>
            </thead>
            <tbody>
  <tr>
    <td class="td-checkbox">
      <input class="form-check-input me-1" type="checkbox" value="apiKey" data-url-param-key="apiKey" data-bundle-name="api key" id="123" checked="" data-can-delete-from-popup="false">
      <label class="form-check-label" for="123">api key</label>
    </td>
    <td>
      <input type="text" class="form-control" data-id="123" placeholder="apiKey value" value="isASecret">
    </td>
    
  </tr>
</tbody>
          </table>
        </div>

        <div class="row">
          <div class="collapse" id="extraParam" style="padding: 0">
            <div class="card card-body">
              <form>
                <div class="col-12">
                  <input required="" type="text" class="form-control" id="bundleName" placeholder="Bundle Name">
                </div>
                <div class="col-12">
                  <input required="" type="text" class="form-control" id="urlParamKey" placeholder="Url param key">
                </div>
                <div class="col-12">
                  <input required="" type="text" class="form-control" id="urlParamValue" placeholder="Url param value">
                </div>
                <div class="col-12">
                  <button id="addNewBundle" style="width: 100%" type="submit" class="btn btn-outline-primary">
                    Add Bundle
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div class="row actions">
          <div class="col-6">
            <button id="applyParams" class="btn btn-outline-primary action-btns">
              Apply Params
            </button>
          </div>
          <div class="col-6">
            <a class="btn btn-outline-success action-btns" data-bs-toggle="collapse" href="#extraParam" role="button" aria-expanded="false" aria-controls="extraParam">
              Add Param
            </a>
          </div>
        </div>
      </div>
    </div>
    <script src="../../libs/bootstrap.min.js"></script>
    <script type="module" src="./index.js"></script>
  

"
`;

exports[`popupTableBuilder should render extra th if an item can be deleted 1`] = `
"

  
    <title>Options</title>

    <link rel="shortcut icon" type="image/jpg" href="../../assets/icons/128x128.png">
    <link href="../../libs/bootstrap.min.css" rel="stylesheet">

    <style>
      .wrapper {
        padding-top: 20px;
        box-sizing: content-box;
      }

      body::-webkit-scrollbar {
        width: 5px;
        height: 8px;
        background-color: #aaa;
      }
      body::-webkit-scrollbar-thumb {
        background: #dc3545;
      }
      #content {
        visibility: hidden;
      }

      #popup-spinner {
        position: absolute;
        left: 45%;
        top: 40%;
      }

      .delete-new-item {
        display: block;
        margin: 0 auto;
      }

      #extraParam .form-control {
        width: 100%;
      }

      #extraParam .col-12 {
        margin: 10px auto;
      }

      .card {
        border: 0;
      }

      .actions {
        padding: 12px 0;
        background-color: rgb(0 0 0 / 90%);
      }

      .action-btns {
        width: 50%;
        margin: 0 auto;
        display: block;
      }

      .table {
        visibility: hidden;
      }

      .td-checkbox {
        display: flex;
        height: 54px;
        align-items: center;
      }
      .td-checkbox label {
        margin-left: 12px;
      }

      #withExtraThead {
        display: none;
      }
    </style>
  
  
    <div class="wrapper">
      <div class="container" id="popup-spinner">
        <div>
          <div class="spinner-border text-primary" role="status"></div>
        </div>
      </div>
      <div class="container-fluid" id="content">
        <div class="row"><h3>QueryParamsBuilder</h3></div>
        <div class="row" style="padding: 0 12px">
          <table class="table table-striped table-bordered">
            <thead>
              <tr>
                <th scope="col" style="display: flex; justify-content: space-between">
                  <span>Params to Apply</span>
                  |
                  <label for="check-all">
                    <input id="check-all" class="form-check-input check-all" type="checkbox">
                    Check All
                  </label>
                </th>
                <th scope="col">Url param value</th>
                <th id="withExtraThead" scope="col">Actions</th>
              </tr>
            </thead>
            <tbody>
  <tr>
    <td class="td-checkbox">
      <input class="form-check-input me-1" type="checkbox" value="apiKey" data-url-param-key="apiKey" data-bundle-name="api key" id="123" checked="" data-can-delete-from-popup="true">
      <label class="form-check-label" for="123">api key</label>
    </td>
    <td>
      <input type="text" class="form-control" data-id="123" placeholder="apiKey value" value="isASecret">
    </td>
    <td><button class="btn btn-danger delete-new-item">-</button></td>
  </tr>
</tbody>
          </table>
        </div>

        <div class="row">
          <div class="collapse" id="extraParam" style="padding: 0">
            <div class="card card-body">
              <form>
                <div class="col-12">
                  <input required="" type="text" class="form-control" id="bundleName" placeholder="Bundle Name">
                </div>
                <div class="col-12">
                  <input required="" type="text" class="form-control" id="urlParamKey" placeholder="Url param key">
                </div>
                <div class="col-12">
                  <input required="" type="text" class="form-control" id="urlParamValue" placeholder="Url param value">
                </div>
                <div class="col-12">
                  <button id="addNewBundle" style="width: 100%" type="submit" class="btn btn-outline-primary">
                    Add Bundle
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div class="row actions">
          <div class="col-6">
            <button id="applyParams" class="btn btn-outline-primary action-btns">
              Apply Params
            </button>
          </div>
          <div class="col-6">
            <a class="btn btn-outline-success action-btns" data-bs-toggle="collapse" href="#extraParam" role="button" aria-expanded="false" aria-controls="extraParam">
              Add Param
            </a>
          </div>
        </div>
      </div>
    </div>
    <script src="../../libs/bootstrap.min.js"></script>
    <script type="module" src="./index.js"></script>
  

"
`;
